### PR TITLE
Implement ingress packet drops alarm

### LINKED ISSUE #965: Implement alarms facility for Snabb and the lwAFTR
Alarm monitoring is a fundamental part of monitoring the network.  Raw alarms from devices do not always tell the status of the network services or necessarily point to the root cause.  However, being able to feed alarms to the network management system in a standardised format is a starting point for performing higher level of network assurance tasks. The ["YANG Alarm Module"](https://tools.ietf.org/html/draft-vallin-netmod-alarm-module-02) draft defines a YANG schema of an alarm interface for network devices.

We will deliver an implementation of an alarms facility in Snabb that follows the capabilities modelled in the YANG alarm module draft.  This facility will provide the following:

 1. Implementation of a `lib.yang.alarms` module in Snabb.

    An alarms module should keep state information of the alarms raised in the system.  It should be usable by any other Snabb program that requires alarm support.  When a forwarding-plane process detects an alarm condition, it will signal that alarm to the main process, which holds an idea of the global state.

    Alarms will be accessible to users via a dedicated `snabb alarms` command.  This command will give users the capability to communicate with the main process of a Snabb network function and perform the actions described as RPCs in the YANG alarms specification:  `compress-alarms`, `purge-alarms`, and `set-operator-state`.  This last command allows users to record alarms as being shelved or resolved.  Users will also be able to list all known alarm types via an `inventory` command.

    Status: Alarms module done (see [alarms.lua](https://github.com/Igalia/snabb/blob/lwaftr/src/lib/yang/alarms.lua)).  `snabb alarms` command implemented also; see generic usage documentation at https://github.com/Igalia/snabb/blob/lwaftr/src/program/alarms/README.md.

 2. Define and implement alarms for the components used in the lwAFTR.  Some planned alarms include:
    
    - [X] Next-hop for IPv4 packets not yet known (ARP)
    - [X] Next-hop for IPv6 packets not yet known (NDP)
    - [X] Ingress bandwidth exceeds 1e9 bytes/s
    - [X] Ingress packet rate exceeds 2 MPPS
    - [X] Ingress packet drop rate exceeds 3333 PPS
    - [X] Bad softwire matches more than 20K softwires/s
    - [X] Incoming ipv4 fragments over 10 KPPS
    - [X] Incoming ipv6 fragments over 10 KPPS
    - [X] Outgoing ipv4 fragments over 10 KPPS
    - [X] Outgoing ipv6 fragments over 10 KPPS
    - [X] Ingress filter v4 rejects over 1 MPPS
    - [X] Ingress filter v6 rejects over 1 MPPS
    - [X] Egress filter v6 rejects over 1 MPPS
    - [X] Egress filter v6 rejects over 1 MPPS

     For specific operational documenation about these alarms, see https://github.com/Igalia/snabb/blob/lwaftr/src/program/lwaftr/doc/alarms.md.

### PR DESCRIPTION
**Pull request by dpino**
This PR implements the following alarm:

* Ingress drop monitor notices ingress packet drop.

Partially fixes #965.

### CODE REVIEW

File: src/lib/timers/ingress_drop_monitor.lua [hunk 1]
```diff
@@ -65,6 +79,8 @@ function IngressDropMonitor:jit_flush_if_needed ()
    if now() - self.last_flush < self.wait then return end
    self.last_flush = now()
    self.last_value[0] = self.current_value[0]
+   counter.set(self.shm['ingress_packet_drop'], self.last_value[0])
```
Comment (wingo) [inline src/lib/timers/ingress_drop_monitor.lua]: nit: use direct access where possible (in this case `self.shm.ingress_packet_drop`)

File: src/lib/timers/ingress_drop_monitor.lua [hunk 2]
```diff
@@ -75,7 +91,22 @@ function IngressDropMonitor:jit_flush_if_needed ()
...
+      ('Ingress packet drop in last interval exceeded limit (%s).'..
+      'This may cause bad performance.'):format(interval)
+   local ingress_packet_drop_alarm = alarms.declare_alarm {
+      [{resource=tostring(S.getpid()),alarm_type_id='ingress-packet-drop'}] = {
+         perceived_severity='major',
+         alarm_text=alarm_text,
+      }
+   }
+   self.ingress_packet_drop_alarm = CounterAlarm.new(ingress_packet_drop_alarm,
+      interval, 1e5, self, 'ingress-packet-drop')
```
Comment (wingo) [inline src/lib/timers/ingress_drop_monitor.lua]: Should this one be `ingress_packet_drop` ?

File: src/lib/timers/ingress_drop_monitor.lua [hunk 3]
```diff
@@ -26,7 +27,10 @@ function new(args)
       tips_url = args.tips_url or default_tips_url,
       last_flush = 0,
       last_value = ffi.new('uint64_t[1]'),
-      current_value = ffi.new('uint64_t[1]')
+      current_value = ffi.new('uint64_t[1]'),
+      shm = {
+         ingress_packet_drop = {counter},
```
Comment (wingo) [inline src/lib/timers/ingress_drop_monitor.lua]: I think this is probably not the right thing; this pattern is designed to be part of an app metatable's `shm` property, and Snabb takes care to `shm.open_frame` this frame.  Whereas here we don't have an app and indeed we have our own counter checking facility.  So maybe remove this `shm = {...}` block?

File: src/lib/timers/ingress_drop_monitor.lua [hunk 4]
```diff
@@ -65,14 +83,21 @@ function IngressDropMonitor:jit_flush_if_needed ()
...
       msg = msg.."; flushing JIT to try to recover"
    end
    msg = msg..". See "..self.tips_url.." for performance tuning tips."
    print(msg)
-   if self.action == 'flush' then jit.flush() end
+
+   self.ingress_packet_drop_alarm:raise({alarm_text=msg})
+   if self.action == 'flush' then
+      jit.flush()
+      self.ingress_packet_drop_alarm:clear()
```
Comment (wingo) [inline src/lib/timers/ingress_drop_monitor.lua]: I think we should clear this alarm in the `if` statement that's on line 82 of this version.  WDYT?

Comment (dpino) [reply src/lib/timers/ingress_drop_monitor.lua]: OK, makes sense.

File: src/lib/timers/ingress_drop_monitor.lua [hunk 5]
```diff
@@ -26,8 +27,9 @@ function new(args)
       tips_url = args.tips_url or default_tips_url,
       last_flush = 0,
       last_value = ffi.new('uint64_t[1]'),
-      current_value = ffi.new('uint64_t[1]')
+      current_value = ffi.new('uint64_t[1]'),
    }
+   ret.threshold = 1
```
Comment (wingo) [inline src/lib/timers/ingress_drop_monitor.lua]: This is from testing I guess; please remove.  Likewise remove the trailing comma on line 30

File: src/lib/timers/ingress_drop_monitor.lua [hunk 6]
```diff
@@ -38,6 +40,20 @@ function new(args)
          ret.counter = counter.open(args.counter)
       end
    end
+
+   alarms.add_to_inventory {
+      [{alarm_type_id='ingress-packet-drop'}] = {
+         resource=tostring(S.getpid()),
+         has_clear=true,
+         description='Ingress packet drops accumulated value',
```
Comment (wingo) [inline src/lib/timers/ingress_drop_monitor.lua]: Not a great description; "Ingress packet drops exceeds threshold" perhaps?

**Review by wingo — APPROVED** [review]

**Pull request merged**